library/core/src/iter/traits/iterator.rs

```diff
@@ -3011,6 +3015,18 @@ pub trait Iterator {
     /// assert_eq!(v_cloned, vec![1, 2, 3]);
     /// assert_eq!(v_map, vec![1, 2, 3]);
     /// ```
+    ///
+    /// To get the best performance, try to clone late:
+    ///
+    /// ```
+    /// let a = ["abc".to_string(), "42".to_string(), ".".to_string()];
+    /// // don't do this:
+    /// let bad: Vec<_> = a.iter().cloned().filter(|s| s.len() == 1).collect();
+    /// assert_eq!(&[".".to_string()], &bad[..]);
+    /// // instead call `cloned` late:
+    /// let good: Vec<_> = a.iter().filter(|s| s.len() == 1).cloned().collect();
+    /// assert_eq!(&[".".to_string()], &good[..]);
+    /// ```
```

Comment:
I think this part should be fairly uncontroversial (modulo readability nits), it simply adds an implementation hint.